apps/emqx_management/src/emqx_mgmt_listeners_conf.erl

```diff
@@ -0,0 +1,126 @@
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_mgmt_listeners_conf).
+
+-behaviour(emqx_config_backup).
+
+-export([
+    action/4,
+    create/3,
+    ensure_remove/2,
+    get_raw/2,
+    update/3
+]).
+
+%% Data backup
+-export([
+    import_data/1
+]).
+
+-include_lib("emqx/include/logger.hrl").
+
+-define(CONF_ROOT_KEY, listeners).
+-define(path(_Type_, _Name_), [?CONF_ROOT_KEY, _Type_, _Name_]).
+-define(OPTS, #{rawconf_with_defaults => true, override_to => cluster}).
+
+action(Type, Name, Action, Conf) ->
+    wrap(emqx_conf:update(?path(Type, Name), {action, Action, Conf}, ?OPTS)).
+
+create(Type, Name, Conf) ->
+    wrap(emqx_conf:update(?path(Type, Name), {create, Conf}, ?OPTS)).
+
+ensure_remove(Type, Name) ->
+    wrap(emqx_conf:tombstone(?path(Type, Name), ?OPTS)).
+
+get_raw(Type, Name) -> emqx_conf:get_raw(?path(Type, Name), undefined).
+
+update(Type, Name, Conf) ->
+    wrap(emqx_conf:update(?path(Type, Name), {update, Conf}, ?OPTS)).
+
+wrap({error, {post_config_update, emqx_listeners, Reason}}) -> {error, Reason};
+wrap({error, {pre_config_update, emqx_listeners, Reason}}) -> {error, Reason};
+wrap({error, Reason}) -> {error, Reason};
+wrap(Ok) -> Ok.
+
+%%------------------------------------------------------------------------------
+%% Data backup
+%%------------------------------------------------------------------------------
+
+import_data(RawConf) ->
+    ListenersConf = maps:get(<<"listeners">>, RawConf, #{}),
+    {Changed, Results} =
+        maps:fold(
+            fun(Type, Listeners, Acc) ->
+                maps:fold(
+                    fun(Name, ListenerConf, {ChangedAcc, ResultsAcc}) ->
+                        {AuthnList, ListenerConf1} = take_authn(ListenerConf),
+                        Path = ?path(Type, Name),
+                        {ChangedAcc1, Action} =
+                            case emqx:get_raw_config(Path, undefined) of
+                                undefined -> {ChangedAcc, create};
+                                _ -> {[Path | ChangedAcc], update}
+                            end,
+                        case emqx_conf:update(Path, {Action, ListenerConf1}, ?OPTS) of
+                            {ok, _} ->
+                                Acc1 = {ChangedAcc1, ResultsAcc#{Path => ok}},
+                                maybe_update_authn(AuthnList, Type, Name, Acc1);
+                            Err ->
+                                {ChangedAcc1, ResultsAcc#{Path => Err}}
+                        end
+                    end,
+                    Acc,
+                    Listeners
+                )
+            end,
+            {[], #{}},
+            ListenersConf
+        ),
+    Info = #{changed => Changed, results => Results, root_key => ?CONF_ROOT_KEY},
+    {ok, Info}.
+
+take_authn(#{<<"authentication">> := Authn} = ListenerConf) ->
```

Comment:
M.b. use`?EMQX_AUTHENTICATION_CONFIG_ROOT_NAME_XXX` as in the other parts of the code?

Reply:
We only have `EMQX_AUTHENTICATION_CONFIG_ROOT_NAME_BINARY` defined in 'emqx/include/emqx_authentication.hrl', but this is the root key (though its name is exactly the same as under listeners).
In general, listeners related code lacks such definitions, for example:
https://github.com/emqx/emqx/blob/master/apps/emqx/src/emqx_listeners.erl#L102
https://github.com/emqx/emqx/blob/master/apps/emqx/src/emqx_listeners.erl#L119

I can introduce new macros, but perhaps we then should do it for all configs.. :thinking: